src/secp256k1.c

```diff
@@ -413,6 +413,80 @@ static SECP256K1_INLINE void buffer_append(unsigned char *buf, unsigned int *off
     *offset += len;
 }
 
+/* Initializes SHA256 with fixed midstate. This midstate was computed by applying
+ * SHA256 to SHA256("BIP340/nonce")||SHA256("BIP340/nonce"). */
+static void secp256k1_nonce_function_bip340_sha256_tagged(secp256k1_sha256 *sha) {
+    secp256k1_sha256_initialize(sha);
+    sha->s[0] = 0xa96e75cbul;
+    sha->s[1] = 0x74f9f0acul;
+    sha->s[2] = 0xc49e3c98ul;
+    sha->s[3] = 0x202f99baul;
+    sha->s[4] = 0x8946a616ul;
+    sha->s[5] = 0x4accf415ul;
+    sha->s[6] = 0x86e335c3ul;
+    sha->s[7] = 0x48d0a072ul;
+
+    sha->bytes = 64;
+}
+
+/* Initializes SHA256 with fixed midstate. This midstate was computed by applying
+ * SHA256 to SHA256("BIP340/aux")||SHA256("BIP340/aux"). */
+static void secp256k1_nonce_function_bip340_sha256_tagged_aux(secp256k1_sha256 *sha) {
+    secp256k1_sha256_initialize(sha);
+    sha->s[0] = 0x5d74a872ul;
+    sha->s[1] = 0xd57064d4ul;
+    sha->s[2] = 0x89495becul;
+    sha->s[3] = 0x910f46f5ul;
+    sha->s[4] = 0xcbc6fd3eul;
+    sha->s[5] = 0xaf05d9d0ul;
+    sha->s[6] = 0xcb781ce6ul;
+    sha->s[7] = 0x062930acul;
+
+    sha->bytes = 64;
+}
+
+static int nonce_function_bip340(unsigned char *nonce32, const unsigned char *msg32, const unsigned char *key32, const unsigned char *xonly_pk32, const unsigned char *algo16, void *data, unsigned int counter) {
+    secp256k1_sha256 sha;
+    unsigned char masked_key[32];
+    int i;
+
+    if (counter != 0) {
+        return 0;
+    }
+    if (algo16 == NULL) {
+        return 0;
+    }
+
+    if (data != NULL) {
```

Comment:
Does this mean that if data  == NULL, we use an unmasked key?

Reply:
Yes. Alternatively we could mask with some static array, but that's unnecessary.

Reply:
I that case we should update the BIP to not do the masking when there is no aux?

Reply:
I think either way is fine. Right now, if you don't provide an aux, you're not BIP compliant which should be clearly stated in the documentation. If you want to be BIP compliant but have no aux, you can use all 0s as the BIP mentions:

> If randomness is not available at all at signing time, a simple counter wide enough to not repeat in practice (e.g., 64 bits or wider) and padded with null bytes to a 32 byte-array can be used, or even the constant array with 32 null bytes.

Reply:
What if we force people to provide an `aux`, i.e., abort if they provide `NULL`?

Reply:
We could also get it if the user randomized the context, but I think we'll then need to increase some PRF counter in the context which will make it not thread safe :(

Reply:
Hm, yes, a counter is a good idea (even if the user hasn't randomized the context). On C11 we could use `<stdatomics.h>` to implement a portable atomic counter but I don't think this is possible on C90.

Reply:
We could use https://gcc.gnu.org/onlinedocs/gcc-4.9.3/gcc/_005f_005fatomic-Builtins.html#g_t_005f_005fatomic-Builtins
and https://docs.microsoft.com/en-us/windows/win32/sync/synchronization-functions?redirectedfrom=MSDN#interlocked-functions

But I don't know what affect will it have on embedded systems.